google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/read_rows/StateMachine.java

```diff
@@ -0,0 +1,414 @@
+          validate(chunk.hasQualifier(), "AWAITING_NEW_ROW: qualifier missing");
+          if (lastCompleteRowKey != null) {
+            validate(
+                ByteStringComparator.INSTANCE.compare(lastCompleteRowKey, chunk.getRowKey()) < 0,
+                "AWAITING_NEW_ROW: key must be strictly increasing");
+          }
+
+          rowKey = chunk.getRowKey();
+          adapter.startRow(rowKey);
+
+          // auto transition
+          return AWAITING_NEW_CELL.handleChunk(chunk);
+        }
+      };
+
+  /** A state that represents a cell boundary in a row. */
+  private final State AWAITING_NEW_CELL =
+      new State() {
+        /**
+         * Processes the next chunk.
+         *
+         * @param chunk The new chunk to process.
+         * @return AWAITING_NEW_CELL if the chunk completed a Cell, AWAITING_CELL_VALUE if the chunk
+         *     left a partial Cell.
+         */
+        @Override
+        State handleChunk(CellChunk chunk) {
+          // Handle row level logic
+          if (chunk.getResetRow()) {
+            return handleResetChunk(chunk);
+          }
+
+          if (!chunk.getRowKey().isEmpty()) {
+            validate(
+                rowKey.equals(chunk.getRowKey()), "AWAITING_NEW_CELL: row keys must not change");
+          }
+
+          // Update cell identifier buffers
+          if (chunk.hasFamilyName()) {
+            familyName = chunk.getFamilyName().getValue();
+            validate(chunk.hasQualifier(), "AWAITING_NEW_CELL: has familyName, but no qualifier");
+          }
+          if (chunk.hasQualifier()) {
+            qualifier = chunk.getQualifier().getValue();
+          }
+          timestamp = chunk.getTimestampMicros();
+          labels = chunk.getLabelsList();
+
+          // Update value expectations
+          validate(chunk.getValueSize() >= 0, "AWAITING_NEW_CELL: valueSize can't be negative");
+
+          boolean isSplit = chunk.getValueSize() > 0;
+          if (isSplit) {
+            validate(
+                !chunk.getCommitRow(),
+                "AWAITING_NEW_CELL: can't commit when valueSize indicates more data");
+            validate(
+                !chunk.getValue().isEmpty(),
+                "AWAITING_NEW_CELL: must have data when valueSize promises more data in the next chunk");
+
+            expectedCellSize = chunk.getValueSize();
+            remainingCellBytes = expectedCellSize - chunk.getValue().size();
+          } else {
+            expectedCellSize = chunk.getValue().size();
+            remainingCellBytes = 0;
+          }
+
+          // Start building cell
+          adapter.startCell(familyName, qualifier, timestamp, labels, expectedCellSize);
+          adapter.cellValue(chunk.getValue());
+
+          // Transitions
+          if (isSplit) {
+            return AWAITING_CELL_VALUE;
+          }
+          adapter.finishCell();
+
+          if (!chunk.getCommitRow()) {
```

Comment:
Maybe inverse the logic?
```
if (chunk.getCommitRow()) {
  return handleCommit;
}

return AWAITING_NEW_CELL;
```

Reply:
I prefer to keep it as is. I tried to keep all transitions in sequential order awaiting for cells happens before committing a row